### ISSUE TITLE
Can't load DerelictSDL2

### ISSUE THREAD

**Issue by SaurabhTotey** [DerelictOrg/DerelictSDL2#64]
I seem to have troubles calling `DerelictSDL2.load();`. This causes the program to exit with code 1 whenever I try compiling it with `dub run --force -v`. My dub.json has `"derelict-sdl2": "~>3.0.0-beta.2"` as a dependency. Even
```
import derelict.sdl2.sdl;
void main() {
    DerelictSDL2.load();
}
```
fails to run. The error I get is
```
derelict.util.exception.SymbolLoadException@../../../../.dub/packages/derelict-util-3.0.0-beta.1/derelict-util/source/derelict/util/sharedlib.d(181): Failed to load symbol SDL_DequeueAudio from shared library libSDL2.so
----------------
../../../../.dub/packages/derelict-util-3.0.0-beta.1/derelict-util/source/derelict/util/sharedlib.d:181 void* derelict.util.sharedlib.SharedLib.loadSymbol(immutable(char)[], bool) [0x445a3e]
../../../../.dub/packages/derelict-util-3.0.0-beta.1/derelict-util/source/derelict/util/loader.d:382 void* derelict.util.loader.SharedLibLoader.loadSymbol(immutable(char)[], bool) [0x443b44]
../../../../.dub/packages/derelict-util-3.0.0-beta.1/derelict-util/source/derelict/util/loader.d:71 void derelict.util.loader.SharedLibLoader.bindFunc(void**, immutable(char)[], bool) [0x443466]
../../../../.dub/packages/derelict-sdl2-3.0.0-beta.2/derelict-sdl2/source/derelict/sdl2/internal/sdl_dynload.d:72 void derelict.sdl2.internal.sdl_dynload.DerelictSDL2Loader.loadSymbols() [0x43fa01]
../../../../.dub/packages/derelict-util-3.0.0-beta.1/derelict-util/source/derelict/util/loader.d:253 void derelict.util.loader.SharedLibLoader.load(immutable(char)[][]) [0x443794]
../../../../.dub/packages/derelict-util-3.0.0-beta.1/derelict-util/source/derelict/util/loader.d:196 void derelict.util.loader.SharedLibLoader.load(immutable(char)[]) [0x443685]
../../../../.dub/packages/derelict-util-3.0.0-beta.1/derelict-util/source/derelict/util/loader.d:134 void derelict.util.loader.SharedLibLoader.load() [0x4434dd]
source/app.d:4 _Dmain [0x43f3da]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFNlZv [0x44a1bf]
??:? scope void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x44a0ef]
??:? scope void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x44a168]
??:? scope void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x44a0ef]
??:? _d_run_main [0x44a05f]
??:? main [0x43f401]
??:? __libc_start_main [0x1bcb882f]
Program exited with code 1
```
which happens regardless of whichever dub run configuration I use. Any help would be appreciated.

**Comment by ZILtoid1991** [comment]
What version of libSDL2.so do you use? I got a similar error once under Windows for using a slightly out of date version of SDL2.dll

**Comment by SaurabhTotey** [comment]
It seems like I have the LibSDL2 version of 2.0.4+dfsg1-2ubuntu2. According to the APT, this was the newest version. I tried running  `sudo apt-get --reinstall install libsdl2-dev`, but that that didn't fix my issues when running the application with dub.

**Comment by ZILtoid1991** [comment]
The symbol SDL_DequeueAudio is present since 2.0.5. You have to somehow get that version, or you're going to have issues with it. Check if your version of Linux have the newest version of LibSDL2

**Comment by SaurabhTotey** [comment]
Huh, that's probably it. Thanks. As of right now, it doesn't seem like there is a LibSDL2 2.0.5 for Linux Mint, so that's unfortunate. Regardless, that seems to be the issue.

**Comment by mdparker** [comment]
> This causes the program to exit with code 1 

That's irrelevant. The interesting information in this case is the message in the first line that starts with "derelict.util.exception.SymbolLoadException". It tells you exactly why the library failed to load. Anytime you're missing a symbol, the first thing to do is to check which version of the shared library you have and which version(s) the binding supports.

In this case, there's an easy workaround. DerelictSDL2 has support for Derelict's `SharedLibVersion`. By default, it will attempt to load SDL 2.0.5, but you can specify a minimum version and it will not throw exceptions when any symbols are missing. So if you know you are never using any functions added later than, say, 2.0.0, you could do this:

```d
DerelictSDL2.load(SharedLibVersion(2, 0, 5);
```

It's all described in the [Derelict documentation](https://derelict.github.io).

**Comment by ZILtoid1991** [comment]
If you want to use 2.0.5, then you can do it by downloading the source of SDL2, then compile it yourself. I did the same with a few applications under Debian, if they don't have a lot of dependencies then it's pretty viable. Also look around Unstable and Testing channels for newer versions. 

Or just force the loader to work with version 2.0.4 or lower.

**Comment by mdparker** [comment]
I strongly recommend using `SharedLibVersion` for anything you plan to distribute. Install what you like on your local machine, but if you aren't using any functions from 2.0.5, there's no reason to require users to have it. Load the minimum version you need.

**Comment by Rageoholic** [comment]
Just would like to note this same bug happens when you specify 2.0.4 and 2.0.3. It looks like there might actually just be a bug here?

**Comment by mdparker** [comment]
Do you mean you're seeing a `SymbolLoadException` that says `SDL_DequeueAudio` is missing when you specify `SharedLibVersion(2,0,4)` or `SharedLibVersion(2,0,3)`?

**Comment by mdparker** [comment]
You're right. There were a couple of problems with the loader. I never caught it because I was always loading SDL 2.0.5. Should be fixed in 3.0.0-beta.4.
